internal/groups/groups.go

```diff
@@ -0,0 +1,112 @@
+package groups
+
+import (
+	"context"
+)
+
+type Member interface{}
+type Metadata map[string]interface{}
+
+type Group struct {
+	ID          string
+	OwnerID     string
+	ParentID    string
+	Name        string
+	Description string
+	Metadata    Metadata
+	// Level if retrieved with Parents or Children method
+	// indicates a level in hierarchy from first group node.
+	Level    int
+	Path     string
+	Children []*Group
+}
+
+type PageMetadata struct {
+	Total  uint64
+	Offset uint64
+	Limit  uint64
+	Name   string
+}
+
+type GroupPage struct {
+	PageMetadata
+	Groups []Group
+}
+
+type MemberPage struct {
+	PageMetadata
+	Members []Member
+}
+
+type Service interface {
+	// CreateGroup creates new  group.
+	CreateGroup(ctx context.Context, token string, g Group) (Group, error)
+
+	// UpdateGroup updates the group identified by the provided ID.
+	UpdateGroup(ctx context.Context, token string, g Group) (Group, error)
+
+	// ViewGroup retrieves data about the group identified by ID.
+	ViewGroup(ctx context.Context, token, id string) (Group, error)
+
+	// ListGroups retrieves groups.
+	ListGroups(ctx context.Context, token string, level uint64, m Metadata) (GroupPage, error)
+
+	// ListChildren retrieves groups that are children to group identified by parentID
+	ListChildren(ctx context.Context, token, parentID string, level uint64, m Metadata) (GroupPage, error)
+
+	// ListParents retrieves groups that are parent to group identified by childID.
+	ListParents(ctx context.Context, token, childID string, level uint64, m Metadata) (GroupPage, error)
+
+	// ListMembers retrieves everything that is assigned to a group identified by groupID.
+	ListMembers(ctx context.Context, token, groupID string, offset, limit uint64, m Metadata) (MemberPage, error)
+
+	// ListMemberships retrieves all groups for member that is identified with memberID belongs to.
+	ListMemberships(ctx context.Context, token, memberID string, offset, limit uint64, m Metadata) (GroupPage, error)
+
+	// RemoveGroup removes the group identified with the provided ID.
+	RemoveGroup(ctx context.Context, token, id string) error
+
+	// Assign adds  member with memberID into the group identified by groupID.
+	Assign(ctx context.Context, token, memberID, groupID string) error
+
+	// Unassign removes member with memberID from group identified by groupID.
+	Unassign(ctx context.Context, token, memberID, groupID string) error
+}
+
+type Repository interface {
+	// Save group
+	Save(ctx context.Context, g Group) (Group, error)
+
+	// Update a group
+	Update(ctx context.Context, g Group) (Group, error)
+
+	// Delete a group
+	Delete(ctx context.Context, groupID string) error
+
+	// RetrieveByID retrieves group by its id
+	RetrieveByID(ctx context.Context, id string) (Group, error)
+
+	// RetrieveByName retrieves group by its name
+	RetrieveByName(ctx context.Context, name string) (Group, error)
```

Comment:
This implies that `Name` is unique. Instead, please consider making (owner, name) pair unique. If two users, for example, group Things by the same criteria (say, spacial), it's very likely that there will be a name collision.

Reply:
Name is unique

Reply:
@mteodor how do you guarantee name uniqueness in a distributed system?

Reply:
Based on the other conversations and on the code in the implementation (`init.go`), the name doesn't seem to be unique. Should this method signature be updated?